### ISSUE TITLE
sysrepod crashed when set item value for bits type

### ISSUE THREAD

**Issue by fanchanghu** [sysrepo/sysrepo#756]
Hi,
it occurs when I use sr_set_item for bits value for second times, it seems this only occurs when the bits value has been set before.

```
*** Error in `sysrepod': double free or corruption (fasttop): 0x00007f08340637b0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x7d1fd)[0x7f084e7391fd]
/nfoam/lib64/libyang.so.0.12(lyd_change_leaf+0x21c)[0x7f084eb0de9e]
/nfoam/lib64/libyang.so.0.12(lyd_new_path+0x25c)[0x7f084eb0eaa5]
/nfoam/lib64/libsysrepo.so(dm_lyd_new_path+0x291)[0x7f084f8e1d08]
/nfoam/lib64/libsysrepo.so(rp_dt_set_item+0x15db)[0x7f084f89d598]
/nfoam/lib64/libsysrepo.so(rp_dt_set_item_wrapper+0x93e)[0x7f084f89fad0]
/nfoam/lib64/libsysrepo.so(+0x9836e)[0x7f084f85c36e]
/nfoam/lib64/libsysrepo.so(+0xb1c15)[0x7f084f875c15]
/nfoam/lib64/libsysrepo.so(+0xb32d0)[0x7f084f8772d0]
/nfoam/lib64/libsysrepo.so(+0xb3d3f)[0x7f084f877d3f]
/lib64/libpthread.so.0(+0x7df5)[0x7f084f3a7df5]
/lib64/libc.so.6(clone+0x6d)[0x7f084e7b21ad]
======= Memory map: ========
00400000-00403000 r-xp 00000000 fd:06 268731191                          /nfoam/bin/sysrepod
00602000-00603000 r--p 00002000 fd:06 268731191                          /nfoam/bin/sysrepod
00603000-00604000 rw-p 00003000 fd:06 268731191                          /nfoam/bin/sysrepod
0203c000-020d8000 rw-p 00000000 00:00 0                                  [heap]
7f0834000000-7f0834078000 rw-p 00000000 00:00 0 
7f0834078000-7f0838000000 ---p 00000000 00:00 0 
7f083c000000-7f083c070000 rw-p 00000000 00:00 0 
7f083c070000-7f0840000000 ---p 00000000 00:00 0 
7f0840000000-7f084004d000 rw-p 00000000 00:00 0 
7f084004d000-7f0844000000 ---p 00000000 00:00 0 
7f0844000000-7f0844089000 rw-p 00000000 00:00 0 
7f0844089000-7f0848000000 ---p 00000000 00:00 0 
7f084b527000-7f084b53c000 r-xp 00000000 fd:06 234881603                  /usr/lib64/libgcc_s-4.8.3-20140911.so.1
7f084b53c000-7f084b73b000 ---p 00015000 fd:06 234881603                  /usr/lib64/libgcc_s-4.8.3-20140911.so.1
7f084b73b000-7f084b73c000 r--p 00014000 fd:06 234881603                  /usr/lib64/libgcc_s-4.8.3-20140911.so.1
7f084b73c000-7f084b73d000 rw-p 00015000 fd:06 234881603                  /usr/lib64/libgcc_s-4.8.3-20140911.so.1
7f084b73d000-7f084b748000 r-xp 00000000 fd:06 234964271                  /usr/lib64/libnss_files-2.17.so
7f084b748000-7f084b947000 ---p 0000b000 fd:06 234964271                  /usr/lib64/libnss_files-2.17.so
7f084b947000-7f084b948000 r--p 0000a000 fd:06 234964271                  /usr/lib64/libnss_files-2.17.so
7f084b948000-7f084b949000 rw-p 0000b000 fd:06 234964271                  /usr/lib64/libnss_files-2.17.so
7f084b949000-7f084b94a000 ---p 00000000 00:00 0 
7f084b94a000-7f084c14a000 rw-p 00000000 00:00 0                          [stack:576]
7f084c14a000-7f084c14b000 ---p 00000000 00:00 0 
7f084c14b000-7f084c94b000 rw-p 00000000 00:00 0                          [stack:575]
7f084c94b000-7f084c94c000 ---p 00000000 00:00 0 
7f084c94c000-7f084d14c000 rw-p 00000000 00:00 0                          [stack:574]
7f084d14c000-7f084d14d000 ---p 00000000 00:00 0 
7f084d14d000-7f084d94d000 rw-p 00000000 00:00 0                          [stack:572]
7f084d94d000-7f084d94e000 r-xp 00000000 fd:06 134288168                  /nfoam/lib64/libyang/zconf.so
7f084d94e000-7f084db4e000 ---p 00001000 fd:06 134288168                  /nfoam/lib64/libyang/zconf.so
7f084db4e000-7f084db4f000 r--p 00001000 fd:06 134288168                  /nfoam/lib64/libyang/zconf.so
7f084db4f000-7f084db50000 rw-p 00002000 fd:06 134288168                  /nfoam/lib64/libyang/zconf.so
7f084db50000-7f084db51000 r-xp 00000000 fd:06 134288167                  /nfoam/lib64/libyang/nacm.so
7f084db51000-7f084dd50000 ---p 00001000 fd:06 134288167                  /nfoam/lib64/libyang/nacm.so
7f084dd50000-7f084dd51000 r--p 00000000 fd:06 134288167                  /nfoam/lib64/libyang/nacm.so
7f084dd51000-7f084dd52000 rw-p 00001000 fd:06 134288167                  /nfoam/lib64/libyang/nacm.so
7f084dd52000-7f084dd53000 r-xp 00000000 fd:06 134288166                  /nfoam/lib64/libyang/metadata.so
7f084dd53000-7f084df53000 ---p 00001000 fd:06 134288166                  /nfoam/lib64/libyang/metadata.so
7f084df53000-7f084df54000 r--p 00001000 fd:06 134288166                  /nfoam/lib64/libyang/metadata.so
7f084df54000-7f084df55000 rw-p 00002000 fd:06 134288166                  /nfoam/lib64/libyang/metadata.so
7f084df55000-7f084dfb4000 r-xp 00000000 fd:06 234997516                  /usr/lib64/libpcre.so.1.2.0
7f084dfb4000-7f084e1b4000 ---p 0005f000 fd:06 234997516                  /usr/lib64/libpcre.so.1.2.0
7f084e1b4000-7f084e1b5000 r--p 0005f000 fd:06 234997516                  /usr/lib64/libpcre.so.1.2.0
7f084e1b5000-7f084e1b6000 rw-p 00060000 fd:06 234997516                  /usr/lib64/libpcre.so.1.2.0
7f084e1b6000-7f084e1b9000 r-xp 00000000 fd:06 234881585                  /usr/lib64/libdl-2.17.so
7f084e1b9000-7f084e3b8000 ---p 00003000 fd:06 234881585                  /usr/lib64/libdl-2.17.so
7f084e3b8000-7f084e3b9000 r--p 00002000 fd:06 234881585                  /usr/lib64/libdl-2.17.so
7f084e3b9000-7f084e3ba000 rw-p 00003000 fd:06 234881585                  /usr/lib64/libdl-2.17.so
7f084e3ba000-7f084e4bb000 r-xp 00000000 fd:06 234964236                  /usr/lib64/libm-2.17.so
7f084e4bb000-7f084e6ba000 ---p 00101000 fd:06 234964236                  /usr/lib64/libm-2.17.so
7f084e6ba000-7f084e6bb000 r--p 00100000 fd:06 234964236                  /usr/lib64/libm-2.17.so
7f084e6bb000-7f084e6bc000 rw-p 00101000 fd:06 234964236                  /usr/lib64/libm-2.17.so
7f084e6bc000-7f084e872000 r-xp 00000000 fd:06 234881561                  /usr/lib64/libc-2.17.so
7f084e872000-7f084ea72000 ---p 001b6000 fd:06 234881561                  /usr/lib64/libc-2.17.so
7f084ea72000-7f084ea76000 r--p 001b6000 fd:06 234881561                  /usr/lib64/libc-2.17.so
7f084ea76000-7f084ea78000 rw-p 001ba000 fd:06 234881561                  /usr/lib64/libc-2.17.so
7f084ea78000-7f084ea7d000 rw-p 00000000 00:00 0 
7f084ea7d000-7f084eb76000 r-xp 00000000 fd:06 100678676                  /nfoam/lib64/libyang.so.0.12.126
7f084eb76000-7f084ed75000 ---p 000f9000 fd:06 100678676                  /nfoam/lib64/libyang.so.0.12.126
7f084ed75000-7f084ed76000 r--p 000f8000 fd:06 100678676                  /nfoam/lib64/libyang.so.0.12.126
7f084ed76000-7f084ed84000 rw-p 000f9000 fd:06 100678676                  /nfoam/lib64/libyang.so.0.12.126
7f084ed84000-7f084ed85000 rw-p 00000000 00:00 0 
7f084ed85000-7f084ed8d000 r-xp 00000000 fd:06 235133741                  /usr/lib64/libprotobuf-c.so.1.0.0
7f084ed8d000-7f084ef8c000 ---p 00008000 fd:06 235133741                  /usr/lib64/libprotobuf-c.so.1.0.0
7f084ef8c000-7f084ef8d000 r--p 00007000 fd:06 235133741                  /usr/lib64/libprotobuf-c.so.1.0.0
7f084ef8d000-7f084ef8e000 rw-p 00008000 fd:06 235133741                  /usr/lib64/libprotobuf-c.so.1.0.0
Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7f084d94c700 (LWP 6534)]
0x00007f084e6f15d7 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install libgcc-4.8.3-9.el7.x86_64
(gdb) bt
#0  0x00007f084e6f15d7 in raise () from /lib64/libc.so.6
#1  0x00007f084e6f2cc8 in abort () from /lib64/libc.so.6
#2  0x00007f084e731e07 in __libc_message () from /lib64/libc.so.6
#3  0x00007f084e7391fd in _int_free () from /lib64/libc.so.6
#4  0x00007f084eb0de9e in lyd_change_leaf (leaf=0x7f0834071a90, val_str=0x7f0844005590 "block")
    at /root/svn/nfoam/libyang/src/tree_data.c:853
#5  0x00007f084eb0eaa5 in lyd_new_path (data_tree=0x7f083405f750, ctx=0x7f083c01c240, 
    path=0x20a8188 "/DataBitsDealer:DataBitsDealer/bitleaf1", value=0x7f0844005590, value_type=LYD_ANYDATA_CONSTSTRING, options=1)
    at /root/svn/nfoam/libyang/src/tree_data.c:1165
#6  0x00007f084f8e1d08 in dm_lyd_new_path (data_info=0x7f0834069c70, path=0x20a8188 "/DataBitsDealer:DataBitsDealer/bitleaf1", 
    value=0x7f0844005590 "block", options=1) at /root/svn/nfoam/sysrepo/src/data_manager.c:6058
#7  0x00007f084f89d598 in rp_dt_set_item (dm_ctx=0x206bd60, session=0x20b8e50, 
    xpath=0x20a8188 "/DataBitsDealer:DataBitsDealer/bitleaf1", options=SR_EDIT_DEFAULT, value=0x20a81b0, str_val=0x0)
    at /root/svn/nfoam/sysrepo/src/rp_dt_edit.c:426
#8  0x00007f084f89fad0 in rp_dt_set_item_wrapper (rp_ctx=0x203c870, session=0x20a6350, 
    xpath=0x20a8188 "/DataBitsDealer:DataBitsDealer/bitleaf1", val=0x20a81b0, str_val=0x0, opt=0)
    at /root/svn/nfoam/sysrepo/src/rp_dt_edit.c:589
#9  0x00007f084f85c36e in rp_set_item_req_process (rp_ctx=0x203c870, session=0x20a6350, msg=0x20a8108)
    at /root/svn/nfoam/sysrepo/src/request_processor.c:1185
#10 0x00007f084f875c15 in rp_req_dispatch (rp_ctx=0x203c870, session=0x20a6350, msg=0x20a8108, skip_msg_cleanup=0x7f084d94be9b)
    at /root/svn/nfoam/sysrepo/src/request_processor.c:3313
#11 0x00007f084f8772d0 in rp_msg_dispatch (rp_ctx=0x203c870, session=0x20a6350, msg=0x20a8108)
    at /root/svn/nfoam/sysrepo/src/request_processor.c:3514
#12 0x00007f084f877d3f in rp_worker_thread_execute (rp_ctx_p=0x203c870) at /root/svn/nfoam/sysrepo/src/request_processor.c:3615
#13 0x00007f084f3a7df5 in start_thread () from /lib64/libpthread.so.0
#14 0x00007f084e7b21ad in clone () from /lib64/libc.so.6
```

**Comment by fabioantonini** [comment]
Hi 
IMHO this bug should have been fixed by the libyang commit bc07ec9476b5e40c11983fade727268ddfe08464 of yesterday?
can you try to confirm that?
fabio

**Comment by fanchanghu** [comment]
Hi, 
It is fixed now, thanks.
